nixos/lib/test-driver/test-driver.py

```diff
@@ -88,6 +90,17 @@
     ")": "shift-0x0B",
 }
 
+# 31-36 inclusive are the main non-greyscale colours. we cycle them infinitely, just in case there are more than 6 machines
+machine_colours_iter = (
+    "\x1b[{}m".format(x) for x in itertools.cycle(reversed(range(31, 37)))
+)
+
+
+@functools.lru_cache()
```

Comment:
i find it a bit unsound that a caching mechanism is used on a function that uses side effects.

How about making the `Machine` constructor pick a color from the iterator and then simply store that for its runtime?

Reply:
long term there is the rough plan to use python's logging frame work. a variant where a logger is injected into each machine's constructor would make a lot of sense, as this can be easily combined with coloring of machines.

Reply:
@tfc it isn't unsound, lru_cache is documented to grow infinitely if no maximum size is specified. Making it in the machine was my original plan, but it gets messy, as the Machine doesn't control the Logger which

Reply:
*which has maybe_prefix, which needs the colour

Reply:
I don't have a strong opinion on that that. The method suggested by @tfc sounds a bit easier to understand is is potentially less code.

Reply:
@Mic92 the serial output doesn't come through maybe_prefix though, it happens directly in Machine.process_serial_output